pkg/policy/api/rules_validation.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+const (
+	maxEntries = 40
+)
+
+type searchRepository func(labels labels.LabelArray) Rules
+
+// ValidateRules performs validation between new and old rules not covered by
+// rule validation.
+func ValidateRules(rules Rules, search searchRepository) error {
+	if err := validateL3(rules, search); err != nil {
+		return err
+	}
+	return nil
+}
+
+// validateL3 checks whether the combined new and old rules bypass the max limit.
+func validateL3(rules Rules, search searchRepository) error {
+	entIngress := make(map[CIDR]bool)
+	entEgress := make(map[CIDR]bool)
+
+	for _, r := range rules {
+		tmp := search(r.Labels)
+		// Add all the previous entries to the map
+		for _, or := range tmp {
+			for _, e := range or.Egress {
+				for _, c := range e.ToCIDR {
+					entEgress[c] = true
+				}
+			}
+			for _, e := range or.Ingress {
+				for _, c := range e.FromCIDR {
+					entIngress[c] = true
+				}
+			}
+		}
+		// Add all the new entries to the map
+		for _, e := range r.Egress {
+			for _, c := range e.ToCIDR {
+				entEgress[c] = true
+			}
+		}
+		for _, e := range r.Ingress {
+			for _, c := range e.FromCIDR {
+				entIngress[c] = true
+			}
+		}
+	}
+
+	if len(entEgress) > maxEntries {
```

Comment:
This function enforces a max 40 CIDR rules across all policy rules. While this avoids compilation failures, it limits the CIDR capability unnecessarily. A better limit is to enforce mac 40 CIDR entries per endpoint. This obviously can't be done at policy validation time, it can only be done when we evaluate the policy for particular endpoints.